feat(lb): Migrate loadbalancer sdk to framework

inpyu opened the pull request:
**Description**

- migrates the existing SDK LoadBalancer and Datasource LoadBalancer to a new framework.
    - Integration of New Framework: The existing SDK LoadBalancer and Datasource LoadBalancer have been migrated to the new framework.

inpyu committed:
feat: migrate loadbalancer sdk to framework

inpyu committed:
feat: migrate loadbalancer sdk to framework

inpyu committed:
chore: add error message

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +41,423 @@ const (
+}
+
+func (l *lbResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = "ncloud"
```
TypeName에 리소스 이름을 지정해주시면 됩니다. ex> `resp.TypeName = req.ProviderTypeName + "_lb"`

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +41,423 @@ const (
-				ForceNew: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
```
framework로 변경되면서 기존 디폴트로 생성되던 id attribute가 필요시 추가하도록 변경되었습니다.
ex) `"id": framework.IDAttribute(),` 추가

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +41,423 @@ const (
-		LoadBalancerNetworkTypeCode: StringPtrOrNil(d.GetOk("network_type")),
-		LoadBalancerName:            StringPtrOrNil(d.GetOk("name")),
-		ThroughputTypeCode:          StringPtrOrNil(d.GetOk("throughput_type")),
+		IdleTimeout:                 ncloud.Int32(int32(plan.IdleTimeout.ValueInt64())),
```
IdelTimeout attritube 가 optional 속성을 가지고 있어서, 입력하지 않는 경우 Null 체크(`plan.IdelTimeout.IsNull()`)를 사전에 추가하면 좋을 것 같습니다. (의도하지 않게 Null인 경우 0으로 요청 인자 설정)

wonchulee reviewed (commented):
안녕하세요! PR 확인이 늦어서 미안합니다.
잠시 짬을 이용해서 전체를 확인해보지 못했지만 몇가지 코멘트 남겨두었습니다.
그리고 테스트는 기존 구sdk를 활용해서 수행하시면 됩니다. (framework의 testing 사용시 의존성 문제로 x) 기존 테스트에서 아주 일부 달라진 부분이 있는데 (provider설정), https://github.com/NaverCloudPlatform/terraform-provider-ncloud/blob/main/internal/service/vpc/vpc_test.go VPC 의 케이스를 참고해보시면 쉽게 파악이 가능하실 것 같습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +41,423 @@ const (
+				"Failed to change throughput type configuration",
+				fmt.Sprintf("Error: %s", err),
+			)
+			return
```
기존 코드에서 throughputType 단일 attribute만 변경시 업데이트가 되지 않던 부분을 함께 개선해주셨네요! 👍 
다만 같은 코드 중복이 발생하다보니, 한 코드 블럭내에서 함께 업데이트할 수 있도록 정리해주시면 좋겠습니다

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +41,423 @@ const (
+		LoadBalancerInstanceNoList: []*string{ncloud.String(state.LoadBalancerNo.ValueString())},
 	}
 
-	if err := waitForLoadBalancerActive(ctx, d, config, d.Id()); err != nil {
-		return diag.FromErr(err)
```
삭제 이전에 LB의 상태 변화가 있다면 active 상태로 변경되는 것을 기다리는 부분이 누락되었는데, 함께 추가해주시면 완벽할 것 같네요

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +41,423 @@ const (
+		default:
+			return retry.NonRetryableError(fmt.Errorf("unexpected load balancer instance operation: %s", operation))
+		}
+	})
```
`state.WaitForStateContext()` 대신 `retry.RetryContext()`를 사용하도록 변경해주셨는데, RetryContext가 내부적으로 WaitForStateContext를 사용하고 있기도하고, 이때 사용되는 StateChangeConf 타입에 상태를 기술하는 것이 아주 살짝.. 가독성이 더 좋아보입니다. 변경하신 이유가 있을 것 같은데, 혹시나 기존처럼 `WaitForStateContext()`를 사용한 형태로 작성 가능하다면 변경해주셔도 좋겠습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +41,423 @@ const (
+func waitFwForLoadBalancerDeletion(ctx context.Context, config *conn.ProviderConfig, id string) error {
+	stateConf := &retry.StateChangeConf{
+		Pending: []string{"PEND"},
+		Target:  []string{"DEL"},
```
"PEND", "DEL" 과 같은 스트링 리터럴 값을 여러번 반복해서 사용하는 경우에, 실수를 방지하기 위해서 기존처럼 const로 정의된 타입을 사용하는 것이 좋습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -302,19 +468,31 @@ func waitForLoadBalancerActive(ctx context.Context, d *schema.ResourceData, conf
+			}
+			if op == "TERMINATING" {
+				return resp, "PEND", nil
+			}
```
status를 체크하는 조건이 기존과 다르게 추가되었는데, 필요한 부분인지 한 번 체크 부탁드립니다.

wonchulee commented on internal/service/loadbalancer/lb_data_source.go:
```diff
@@ -2,60 +2,166 @@ package loadbalancer
 		},
-		"filter": DataSourceFiltersSchema(),
 	}
-	return GetSingularDataSourceItemSchemaContext(ResourceNcloudLb(), fieldMap, dataSourceNcloudLbRead)
```
기존 코드에서 `GetSingularDataSourceItemSchemaContext(ResourceNcloudLb(), fieldMap, dataSourceNcloudLbRead)` 함수를 사용해 lb Resource에 정의된 attribute들을 data source에 그대로 적용(중복된 것 제외)하도록 사용하고 있었는데요, 현재 framework에서도 동일한 역할을 하는 helper 함수를 만들어 제공하고 있지 않습니다. 해당 helper 함수를 작성하거나, 기존 Resource에 정의되어 있는 attribute들을 직접 data source에 정의해주는 것이 필요합니다.

wonchulee commented on internal/service/loadbalancer/lb_data_source.go:
```diff
@@ -2,60 +2,166 @@ package loadbalancer
+	return lbList, diags
 }
 
 func getVpcLoadBalancerList(config *conn.ProviderConfig, id string) ([]*LoadBalancerInstance, error) {
```
getVpcLoadBalancerList()를 사용하지 않도록 변경되어, 삭제해주셔도 좋겠습니다.

youngmn commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +41,423 @@ const (
+		return
 	}
 
-	throughput_type := StringPtrOrNil(d.GetOk("throughput_type"))
-	if (d.Get("type").(string) == "NETWORK") && (throughput_type != nil && *throughput_type != "DYNAMIC") {
-		return diag.FromErr(fmt.Errorf("Network Loadbalancer throughput_type can only be set to empty or DYNAMIC"))
```
https://github.com/NaverCloudPlatform/terraform-provider-ncloud/pull/427
여기서 반영된 내용인데, 누락된 것으로 보입니다.

inpyu committed:
chore: lb 리소스 이름 지정

inpyu committed:
fix: idleTimeoutValue null값 개선

inpyu committed:
fix: 코드 중복 개선

inpyu committed:
fix: delete시 누락된 active 코드 수정

inpyu committed:
fix: 코드 리뷰 반영 및 각종 오류 개선

inpyu commented:
> 안녕하세요! PR 확인이 늦어서 미안합니다. 잠시 짬을 이용해서 전체를 확인해보지 못했지만 몇가지 코멘트 남겨두었습니다. 그리고 테스트는 기존 구sdk를 활용해서 수행하시면 됩니다. (framework의 testing 사용시 의존성 문제로 x) 기존 테스트에서 아주 일부 달라진 부분이 있는데 (provider설정), https://github.com/NaverCloudPlatform/terraform-provider-ncloud/blob/main/internal/service/vpc/vpc_test.go VPC 의 케이스를 참고해보시면 쉽게 파악이 가능하실 것 같습니다.

리뷰 감사합니다! 해당 리뷰 내용 반영해서 수정 진행하였고, 테스트 결과는 다음과 같습니다.
```
=== RUN   TestAccResourceNcloudLb_vpc_basic
=== PAUSE TestAccResourceNcloudLb_vpc_basic
=== CONT  TestAccResourceNcloudLb_vpc_basic
    d:\programming\terraform-provider-ncloud\internal\service\loadbalancer\lb_test.go:22: Acceptance tests skipped unless env 'TF_ACC' set
--- SKIP: TestAccResourceNcloudLb_vpc_basic (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-ncloud/internal/service/loadbalancer  0.992s
```

```
=== RUN   TestAccDataSourceNcloudLb_basic
=== PAUSE TestAccDataSourceNcloudLb_basic
=== CONT  TestAccDataSourceNcloudLb_basic
    d:\programming\terraform-provider-ncloud\internal\service\loadbalancer\lb_data_source_test.go:17: Acceptance tests skipped unless env 'TF_ACC' set
--- SKIP: TestAccDataSourceNcloudLb_basic (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-ncloud/internal/service/loadbalancer  0.282s

```

dusdjhyeon committed:
refactor: change msg datasource to resource

dusdjhyeon committed:
fix: change valid length (#434)

dusdjhyeon committed:
fix(autoscaling): Autoscaling test failure (#436)

* fix: autoscalig acctest

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -339,6 +499,36 @@ func GetVpcLoadBalancer(config *conn.ProviderConfig, id string) (*LoadBalancerIn
 }
 
+func (l *lbResourceModel) refreshFromOutput(ctx context.Context, output *LoadBalancerInstance) {
+	l.LoadBalancerNo = types.StringPointerValue(output.LoadBalancerInstanceNo)
```
새로 추가하신 `id` attribute 의 값도 LoadBalancerInstanceNo 값으로 동일하게 설정이 필요합니다.
LoadBalancerNo와 중복된 값을 사용하고 있는데, 이전 리소스와 동일하게 호환성을 제공하기 위해 필요한 부분입니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -339,6 +499,36 @@ func GetVpcLoadBalancer(config *conn.ProviderConfig, id string) (*LoadBalancerIn
+	for _, listener := range output.LoadBalancerListenerList {
+		listenerNoList = append(listenerNoList, *listener)
+	}
+	l.ListenerNoList, _ = types.ListValueFrom(ctx, types.StringType, listenerNoList)
```
`types.ListValueFrom()` 가 실패하는 경우 전달되는 Diagnostics 값을 처리하도록 변경해주시면 미래의 실수를 예방할 수 있어서 더 좋을 것 같습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -355,3 +545,18 @@ func convertVpcLoadBalancer(instance *vloadbalancer.LoadBalancerInstance) *LoadB
 	}
 }
+
+type lbResourceModel struct {
```
위에서 언급한 것처럼 Model에 id attribute에 해당하는 값 추가가 필요합니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -26,272 +40,417 @@ const (
-		Importer: &schema.ResourceImporter{
-			StateContext: schema.ImportStatePassthroughContext,
-		},
-		Timeouts: &schema.ResourceTimeout{
-			Create: schema.DefaultTimeout(conn.DefaultCreateTimeout),
-			Update: schema.DefaultTimeout(conn.DefaultUpdateTimeout),
-			Delete: schema.DefaultTimeout(conn.DefaultTimeout),
-		},
```
기존 리소스 호환성 유지를 위해서 framework에서도 timeouts 정의가 필요합니다.
https://developer.hashicorp.com/terraform/plugin/framework/migrating/resources/timeouts 해당 내용 참고하여 timeout 설정 추가 부탁드립니다~

inpyu committed:
feat: 기존 리소스와의 호환을 위한 ID값 추가

inpyu committed:
feat: types.ListValueFrom 에러처리 추가

inpyu committed:
feat: 호환성을 위한 timeout 설정 추가

inpyu committed:
git rebase

inpyu commented:
안녕하세요 ! @wonchulee  추가적으로 피드백 주신 내용 반영하여 수정했습니다! 다만 Timeout 설정을 하는 과정에서 Timeout의 도입으로 go mod의 버전이 변경되었는데 이부분이 바뀌어도 괜찮은 걸까요 ..?

Geun-Oh committed:
chore: Upgrade Go version to v1.21

chore: Upgrade Go version to v1.21

chore: upgrade Go to v1.21

ci: Set go version 1.21 in GitHub Actions

ci: Upgrade actions version (checkout, setup-go, ci-lint)

ci: Upgrade golangci-lint v1.59

Geun-Oh committed:
feat: Add Object Storage service (#439)

* chore: Upgrade Go version to v1.21

chore: Upgrade Go version to v1.21

chore: upgrade Go to v1.21

ci: Set go version 1.21 in GitHub Actions

ci: Upgrade actions version (checkout, setup-go, ci-lint)

ci: Upgrade golangci-lint v1.59

* chore: Install AWS SDK for Object Storage API

* feat: Add S3 Client configuration

* feat: Add Object Storage Resources

* feat: Add objectstorage_bucket resource

* feat: Add objectstorage_bucket_acl resource

* feat: Add objectstorage_object resource

* feat: Add objectstorage_object_acl resource

* fix: add "_bucket" suffix to bucket resource

* feat: Add examples of object storage

* docs: Add initial documentation of object storage

* feat: add type casting function for nested listValue

* feat: Add BucketRegion to bucket chema

* fix: Add detailed schema of grants & type casting method

* chore: Add error exception

* feat: Add regex validator for resource id

* chore: remove redundant EOL

* fix: Remove BucketName, BucketRegion from schema(doesn't provided)

* refactor: Add TrimForIDParsing util

* feat: Add temporary file creation logic

* feat: Add aclOptions for testing

* fix: Removed Unavailable object acl

* feat: Add Bucket data source

* fix: Add new bucket creation for independant testing

* fix: aws.String => ncloud.String

* chore: fix typo

* chore: change string value with constants

* refactor: expand TrimForParsing to parse vairous strings (not only ID)

* feat: use region info from initial configuration

* refactor: Set bucket id same as bucket name

* fix: Add detailed string validator with regex for bucket_name

* feat: Add NewObjectDataSource data source

* refactor: Move ObjectIDParser to object.go

* refactor: Set Object Resource ID with new format

* feat: set independent object testing

* feat: Add Owner info & Creation Date to bucket data source

* feat: Add OwnerID & OwnerDisplayname at object_acl attribute

* fix: set bucket_id & object_id w/o NRN

* docs: Add documentation of object storage resources

* feat: Change format - make resource id with '/' in object storage - object

* feat: Add objectstorage_object_copy

* fix: Change to do not use pointer method

* docs: Add data source documentation of bucket

* docs: Add object_storage_object.md in data-source

* docs: Add object_storage_object_copy.md

* feat: Add CreationDate attribute to bucket

* fix: fix logic to check whether output is nil

* fix: Add format of creationDate

* feat: Delete panic from object_copy Update

* feat: Remove endpoint from Config struct

* fix: Remove Optional keyword from creation_date(computed)

* refactor: set bucket unique value from bucket_id to bucket_name

* fix: Remove unused headbucket

* feat: Add bucket ID (since it uses id for default)

* feat: Add attributes based on HeadObjectOutput

* feat: Add endpoint abstract logic

* refactor: Move post fetch logic into refreshFromOutput

* feat: Add genEndpointWithCode (region, site)

* feat: Add optional endpoint configuration

* feat: Add region formatting logic

* feat: Add Optional variables compatible with aws sdk in object-kind resource

* docs: Update docs with additional attributes

* fix: Remove Object Encryption logic entirely

* feat: Add site variable in example tf code

* chore: Add Error with Diagnostics

* refactor: remove endpoint from providerConfig

* chore: add missed packages

* fix: fix endpoint env to optional

* fix: Parse bucket info after pulling plan data

* refactor: Abstract BucketNameValidator

* fix: Remove optional from creation_date

* fix: Change serverless environment => platform independent

* fix: change id with object_url

* feat: Add attribute documentation

* feat: Add update logic to object_acl

* feat: Add update logic to bucket_acl

* fix: Fix typo

Revert "fix: Fix typo"

This reverts commit 37c773743b0f01e70fd675a16f3b77592117dcf2.

fix: Fix typo

dusdjhyeon committed:
[Feat] Add Login Key Data Source (#442)

* feat: add login key data source

* docs: add login key docs

soosue committed:
docs: Fix typo (#438)

* docs(redis): Correct indentation in main.if

* docs(server): Correct indentation in server_products.md

* docs(mysql): Fix typo in mysql_products.md

* docs(mssql): Fix typo in mssql_products.md

wonchulee commented on internal/service/loadbalancer/lb_data_source.go:
```diff
@@ -2,80 +2,233 @@ package loadbalancer
+		"subnet_no_list": schema.ListAttribute{
+			ElementType: types.StringType,
+			Required:    true,
+		},
```
새롭게 추가해주신 `network_type`, `idle_timeout`, `type`, `throughput_type`, 그리고 `description`은 사용자가 입력한 값을 사용하지 않고, lb의 id값으로 가져온 값을 사용할 수 있도록 제공하기 때문에 computed 속성만 갖는 attribute로 정의하는 것이 필요합니다. 
그리고 아래 flattenLoadBalancers()과 loadBalancerDataSourceModel 에 추가된 attribute 값을 함께 정의 및 값 설정하도록 추가가 필요합니다.

wonchulee reviewed (commented):
매번 리뷰가 늦어서 미안합니다. 다른 PR이 반영되어 main 브랜치 기준으로 리베이스 부탁드려요, 그리고 코멘트 남긴 것만 개선해주시면 그 외엔 다른 문제 없어보입니다. 👍

wonchulee commented:
> 안녕하세요 ! @wonchulee 추가적으로 피드백 주신 내용 반영하여 수정했습니다! 다만 Timeout 설정을 하는 과정에서 Timeout의 도입으로 go mod의 버전이 변경되었는데 이부분이 바뀌어도 괜찮은 걸까요 ..?

네, 작업해주신 변경 사항에서 문제될 부분 없습니다!

Geun-Oh committed:
feat: Add documentation about object storage endpoint (#443)

* feat: Add util functions for endpoint formatting

feat: Add documentation about object storage endpoint

feat: Add mapRegionWithEnv util func

docs: Add env configuration guide

feat: remove endpoint from docs

feat: Modify endpoint extracting func

* docs: Add extra comment for optional env

inpyu committed:
fix: go mod tidy 수정

inpyu committed:
feat: lb attribute 추가

inpyu committed:
Merge branch 'main' into feature/lb_fw

inpyu committed:
fix: go mod 버전 conflict 수정

inpyu committed:
fix: 사용되지 않는 모듈 삭제

inpyu commented:
분명 rebase를 깔끔하게 하려고 했는데 이상하게 커밋이 안합쳐진 듯 하네요 ㅠㅠ 우선 코멘트 해주신 부분 개선했습니다.. ! @wonchulee

wonchulee committed:
fix(lb): Fix migration Errors

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -81,6 +76,9 @@ func (l *lbResource) Schema(ctx context.Context, req resource.SchemaRequest, res
 			},
 			"domain": schema.StringAttribute{
 				Computed: true,
```
기존 lb 사용자에게 리그레이션을 발생시킬 수 있는 부분이라서 Computed 속성의 attribute들에게 UseStateForUnknown() planmodifier를 추가했습니다. 변경사항이 없음에도 Computed 값으로 인해 update-in-place 동작이 발생하는 것을 막기 위해서 추가했습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -92,11 +90,14 @@ func (l *lbResource) Schema(ctx context.Context, req resource.SchemaRequest, res
 				},
 			},
-			"idle_timeout": schema.Int64Attribute{
+			"idle_timeout": schema.Int32Attribute{
```
Int32 타입으로 변경했습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -128,15 +135,23 @@ func (l *lbResource) Schema(ctx context.Context, req resource.SchemaRequest, res
 			"timeouts": timeouts.Block(ctx, timeouts.Opts{
 				Create: true,
+				Update: true,
+				Delete: true,
```
기존 코드 베이스에 update, delete 도 함께 정의되어 있어서 추가했습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -187,32 +202,30 @@ func (l *lbResource) Create(ctx context.Context, req resource.CreateRequest, res
+
+	if !plan.IdleTimeout.IsNull() && !plan.IdleTimeout.IsUnknown() {
+		reqParams.IdleTimeout = plan.IdleTimeout.ValueInt32Pointer()
+	}
```
throughputType과 idleTimeout 값을 입력하지 않는 경우에 빈 값("", 0) 으로 설정되어 의도한 것과 다르게 동작하는 부분이 있었습니다. 값이 있는 경우에만 요청 파라미터로 설정하게 변경했습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -342,7 +355,7 @@ func (l *lbResource) Update(ctx context.Context, req resource.UpdateRequest, res
 		!plan.Description.Equal(state.Description) {
 
-		if err := waitForLoadBalancerActive(ctx, l.config, state.LoadBalancerNo.String()); err != nil {
+		if err := waitForLoadBalancerActive(ctx, l.config, state.LoadBalancerNo.ValueString()); err != nil {
```
`String()` 을 사용하여 " quote 값까지 포함되어 상태 변화 체크가 정상적으로 되지 않는 부분을 ValueString을 사용하도록 변경했습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -354,25 +367,44 @@ func (l *lbResource) Update(ctx context.Context, req resource.UpdateRequest, res
+			}
+
+			state.Description = plan.Description
+		}
```
description, IdleTimeout/ThroughputTypeCode 값 변경을 한 번에 요청할 수 있는 경우가 있기 때문에 if/else 조건에서 각각의 if 조건으로 변경했습니다. 그리고 각 요청 사이에 lb의 상태변화가 있기 때문에 기존코드처럼 active 상태 변화를 기다리는 부분을 다시 추가했습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -500,10 +521,11 @@ func waitForLoadBalancerDeletion(ctx context.Context, config *conn.ProviderConfi
 
-			return nil, "", fmt.Errorf("error occurred while waiting to delete load balancer instance")
+			respCode := resp.LoadBalancerInstanceList[0].LoadBalancerInstanceOperation.Code
+			return nil, ncloud.StringValue(respCode), nil
```
삭제 상태를 체크하는 조건이 변경되었는데, 정상적으로 삭제 상태를 체크하지 못하고 있었습니다. 기존 로직으로 변경했습니다.

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -443,17 +475,6 @@ func GetFwVpcLoadBalancer(ctx context.Context, config *conn.ProviderConfig, id s
-		}
-	}
-	return result
-}
```
미사용 코드 삭제

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -542,14 +564,17 @@ func (l *lbResourceModel) refreshFromOutput(ctx context.Context, output *LoadBal
 	l.ThroughputType = types.StringPointerValue(output.ThroughputType)
 	l.VpcNo = types.StringPointerValue(output.VpcNo)
 
+	if output.LoadBalancerDescription != nil && *output.LoadBalancerDescription != "" {
+		l.Description = types.StringPointerValue(output.LoadBalancerDescription)
+	}
```
기존 구SDK와 호환성을 위해 Description이 없는 경우 nil 값으로 설정했습니다. (framework 코드에서 "" 빈값으로 설정하여 diff 발생)

wonchulee commented on internal/service/loadbalancer/lb_data_source.go:
```diff
@@ -71,7 +69,7 @@ var lbResourceSchema = schema.Schema{
 		"subnet_no_list": schema.ListAttribute{
 			ElementType: types.StringType,
-			Required:    true,
+			Computed:    true,
```
computed 속성 변경 누락된 부분이 있어서 변경했습니다.

wonchulee commented on internal/service/loadbalancer/lb_data_source.go:
```diff
@@ -230,25 +228,18 @@ func flattenLoadBalancers(ctx context.Context, list []*vloadbalancer.LoadBalance
-	SubnetNoList   types.List     `tfsdk:"subnet_no_list"`
-	IpList         types.List     `tfsdk:"ip_list"`
-	ListenerNoList types.List     `tfsdk:"listener_no_list"`
-	Timeouts       timeouts.Value `tfsdk:"timeouts"`
```
불필요한 Timeouts 필드가 있어 삭제했습니다.

wonchulee committed:
fix: Add idleTimeout validator

wonchulee commented on internal/service/loadbalancer/lb.go:
```diff
@@ -28,270 +42,476 @@ const (
 				Computed: true,
+				Validators: []validator.Int32{
+					int32validator.Between(1, 3600),
+					checkUnsupportedIdleTimeout("NETWORK"),
```
리그레션이 있나 확인하던 중 발견한 부분인데, network type의 lb는 `idle_timeout` 값 설정이 불가하군요,
이 경우 사용자가 명시적으로 idle_timeout 값을 설정하더라도 응답에는 해당 값이 제대로 설정되지 않아 tf 상태 불일치가 발생했습니다. 기존 구SDK 사용시 sdk의 결함으로 무시될 수 있어서 문제되지 않았었는데, framework로 변경 후 상태 불일치가 확인되어 문제가 있을 수 있습니다.
기존 동작과 달라지는 부분이지만, 명시적으로 netowork type의 lb는 idel_timeout 값을 설정할 수 없도록 custom validator를 추가하도록 변경했습니다. (참고로 코멘트 남깁니다. cc @youngmn)

wonchulee reviewed (commented):
안녕하세요, 변경해주신 내용 확인하다보니, 이것저것 수정하게 되어 변경사항 함께 올려두었습니다. (참고하시라고 변경의 이유 코멘트로 함께 남겨두었습니다)
기존 lb를 사용하던 고객들이 리팩토링된 lb 버전으로도 문제없이 사용할 수 있어야 하기에, 기존 리소스와 번갈아가며 동작을 검증했습니다.
기존 테스트로는 커버되지 않는 경우가 많아서 tf configuration 코드를 직접 작성하여 사용자가 입력할 수 있는 여러 경우의 수 테스트가 필요한데요
- 구 SDK로 생성한 리소스를 리팩토링 된 terraform provider 버전으로 동작 검증
- 신규 SDK로 리소스 생성/업데이트/삭제

위 케이스에 대해서 한 번 크로스 체크로 실행해서 확인해주시고, 반영하면 될 것 같습니다! 확인 후 코멘트 남겨주세요. 고생하셨습니다.

wonchulee committed:
fix(lb): Handle error properly

wonchulee committed:
fix: Generalize confictsWithValue validators

wonchulee committed:
fix: Generalize conflictsWithValue validator again

inpyu commented:
> 안녕하세요, 변경해주신 내용 확인하다보니, 이것저것 수정하게 되어 변경사항 함께 올려두었습니다. (참고하시라고 변경의 이유 코멘트로 함께 남겨두었습니다) 기존 lb를 사용하던 고객들이 리팩토링된 lb 버전으로도 문제없이 사용할 수 있어야 하기에, 기존 리소스와 번갈아가며 동작을 검증했습니다. 기존 테스트로는 커버되지 않는 경우가 많아서 tf configuration 코드를 직접 작성하여 사용자가 입력할 수 있는 여러 경우의 수 테스트가 필요한데요
> 
> * 구 SDK로 생성한 리소스를 리팩토링 된 terraform provider 버전으로 동작 검증
> * 신규 SDK로 리소스 생성/업데이트/삭제
> 
> 위 케이스에 대해서 한 번 크로스 체크로 실행해서 확인해주시고, 반영하면 될 것 같습니다! 확인 후 코멘트 남겨주세요. 고생하셨습니다.

모두 확인했습니다!! 감사드립니다 :)
